pkg/kclient/all.go

```diff
@@ -59,15 +61,18 @@ func getAllResources(client dynamic.Interface, apis []apiResource, ns string, se
 	klog.V(2).Infof("fired up all goroutines to query APIs")
 
 	go func() {
-		for val := range outChan {
-			out = append(out, val...)
+		for {
+			select {
+			case <-errChan:
+				return
+			case val := <-outChan:
+				out = append(out, val...)
```

Comment:
> I certainly miss something about goroutines, but does that mean that this goroutine will keep running indefinitely if there was nothing written to `errChan`? I don't see how it would return from this infinite loop otherwise. Again, as we previously discussed, I am thinking about long-running processes like `odo logs --follow`..

I'm struggling really badly with imagining things from `odo logs --follow` perspective. 

> But it seems to me that the channel was closed after all the goroutines returned. So I am not sure I understand why the channel would get closed too soon.

I initially ignored few test failures because they didn't happen in `odo logs` or `odo list`. But one failure in `odo delete component` piqued my interest. In the logs, it shows that 1 resource of kind `Deployment` was found, but the message printed when we do `odo delete component -f` didn't include the `Deployment: <name of the deployment>`. My understanding is that this was because of the channel getting closed just a little too early. Also, this failure was intermittent and did not happen 100% of the times. For example, this morning when I cancelled a few builds of Charlie's PR, the CI passed just fine with the previous commit.

> Or would the approach[1] of iterating over `outChan` in the `getAllResources` function itself (i.e., not in a goroutine) be helpful, in the sense that the `for-range` loop would block until `outChan` is closed? So we would be sure that we are done iterating over `outChan` before the function returns, while not leaking any goroutines in this function.

I did think about iterating over `outChan` without a goroutine. But where would I put the `group.Wait()` in that case? I think that:
*  if we put it after the loop that's iterating over `outChan`, it won't let `group.Wait()` return an error as soon as one of the goroutines returns an error. 
* if we put `group.Wait()` before the iteration over `outChan`, we would have to make `outChan` buffered to the size of `apisOfInterest`.

So I'm considering doing `defer close(errChan)` and `defer close(outChan)` right after creating them. So, when `getAllResources` is done, it would close the channels and the goroutine would no longer be waiting on the two channels. WDYT?

Reply:
> > I certainly miss something about goroutines, but does that mean that this goroutine will keep running indefinitely if there was nothing written to `errChan`? I don't see how it would return from this infinite loop otherwise. Again, as we previously discussed, I am thinking about long-running processes like `odo logs --follow`..
> 
> I'm struggling really badly with imagining things from `odo logs --follow` perspective.

I mean, I am just not able to keep `odo logs --follow` in my head while coding on this PR.

Reply:
> But one failure in `odo delete component` piqued my interest. In the logs, it shows that 1 resource of kind `Deployment` was found, but the message printed when we do `odo delete component -f` didn't include the `Deployment: <name of the deployment>`. My understanding is that this was because of the channel getting closed just a little too early. Also, this failure was intermittent and did not happen 100% of the times.

Just faced this in the latest failure. Here's the excerpt:
```
<< End Captured GinkgoWriter Output

  Expected
      <string>: Searching resources to delete, please wait...
      This will delete "mynodejs" from the namespace "cmd-delete-test186eqw".
       •  The component contains the following resources that will get deleted:
      	- Service: my-cs
      	- Endpoints: my-cs
      The component "mynodejs" is successfully deleted from namespace "cmd-delete-test186eqw"
      
  to contain substring
      <string>: Deployment: my-component
  In [It] at: /go/odo_1/tests/integration/cmd_delete_test.go:188
```
I'm probably going to get rid of goroutines, or maybe just revert the `sync.Mutex` that was there earlier.

Reply:
> > I certainly miss something about goroutines, but does that mean that this goroutine will keep running indefinitely if there was nothing written to `errChan`? I don't see how it would return from this infinite loop otherwise. Again, as we previously discussed, I am thinking about long-running processes like `odo logs --follow`..
> 
> I'm struggling really badly with imagining things from `odo logs --follow` perspective.
> 
> > But it seems to me that the channel was closed after all the goroutines returned. So I am not sure I understand why the channel would get closed too soon.
> 
> I initially ignored few test failures because they didn't happen in `odo logs` or `odo list`. But one failure in `odo delete component` piqued my interest. In the logs, it shows that 1 resource of kind `Deployment` was found, but the message printed when we do `odo delete component -f` didn't include the `Deployment: <name of the deployment>`. My understanding is that this was because of the channel getting closed just a little too early. Also, this failure was intermittent and did not happen 100% of the times. For example, this morning when I cancelled a few builds of Charlie's PR, the CI passed just fine with the previous commit.
> 
> > Or would the approach[1] of iterating over `outChan` in the `getAllResources` function itself (i.e., not in a goroutine) be helpful, in the sense that the `for-range` loop would block until `outChan` is closed? So we would be sure that we are done iterating over `outChan` before the function returns, while not leaking any goroutines in this function.
> 
> I did think about iterating over `outChan` without a goroutine. But where would I put the `group.Wait()` in that case? I think that:
> 
> * if we put it after the loop that's iterating over `outChan`, it won't let `group.Wait()` return an error as soon as one of the goroutines returns an error.
> * if we put `group.Wait()` before the iteration over `outChan`, we would have to make `outChan` buffered to the size of `apisOfInterest`.
> 
>

Indeed, I agree with these points; that's why I suggested we put both `group.Wait()` and `close(outChan)` in the same "monitoring" goroutine? Well, that's a pattern I saw used with `sync.WaitGroup` (see for example the realistic example depicted in the "Using WaitGroups" section of the "Learning Go" book - pages 221 through 222).

And if we don't want to make `errChan` buffered, I think writing to it after closing `outChan` (rather than before) should fix the blocking issue. The write to `errChan` would block until a read happens, which will actually happen because the for-range loop was done (because `outChan`  was closed). Like so:

```go
    klog.V(2).Infof("fired up all goroutines to query APIs")

	errChan := make(chan error)
	go func() {
		err := group.Wait()
		klog.V(2).Infof("all goroutines have returned in %v", time.Since(start))
		close(outChan)
		errChan <- err
	}()

	for v := range outChan {
		out = append(out, v...)
	}

	klog.V(2).Infof("query result: objects=%d", len(out))
	return out, <-errChan
```

Reply:
> > But one failure in `odo delete component` piqued my interest. In the logs, it shows that 1 resource of kind `Deployment` was found, but the message printed when we do `odo delete component -f` didn't include the `Deployment: <name of the deployment>`. My understanding is that this was because of the channel getting closed just a little too early. Also, this failure was intermittent and did not happen 100% of the times.
> 
> Just faced this in the latest failure. Here's the excerpt:
> 
> ```
> << End Captured GinkgoWriter Output
> 
>   Expected
>       <string>: Searching resources to delete, please wait...
>       This will delete "mynodejs" from the namespace "cmd-delete-test186eqw".
>        •  The component contains the following resources that will get deleted:
>       	- Service: my-cs
>       	- Endpoints: my-cs
>       The component "mynodejs" is successfully deleted from namespace "cmd-delete-test186eqw"
>       
>   to contain substring
>       <string>: Deployment: my-component
>   In [It] at: /go/odo_1/tests/integration/cmd_delete_test.go:188
> ```
> 
> I'm probably going to get rid of goroutines, or maybe just revert the `sync.Mutex` that was there earlier.
> 
Can we try the approach I suggested above in https://github.com/redhat-developer/odo/pull/5973#discussion_r937164777 and see how it goes?

Reply:
> Can we try the approach I suggested above in [#5973 (comment)](https://github.com/redhat-developer/odo/pull/5973#discussion_r937164777) and see how it goes?

Sure. By just reading the code, I think it should solve things for us.

Reply:
Done. Let's see if the CI gods like it. As a relief to me, the k8s IBM Cloud tests and OCP prow tests both had the same failure. OCP IBM Cloud was showing different errors, but I'm not bothered about those at the moment. Let's see.

Reply:
Thanks Dharmit.